node/engine/messageservice/p2p-message-service/service.go

```diff
@@ -83,71 +93,123 @@ func NewMessageService(ip string, port int, me types.Address, pk []byte, useMdns
 	options := []libp2p.Option{
 		libp2p.Identity(messageKey),
 		libp2p.ListenAddrStrings(fmt.Sprintf("/ip4/%s/tcp/%d", ip, port)),
-		libp2p.Transport(tcp.NewTCPTransport),
-		libp2p.NoSecurity,
-		libp2p.DefaultMuxers,
+		// libp2p.Transport(tcp.NewTCPTransport),
+		// libp2p.NoSecurity,
+		// libp2p.DefaultMuxers,
 	}
 	host, err := libp2p.New(options...)
 	if err != nil {
 		panic(err)
 	}
 
 	ms.p2pHost = host
-
 	ms.p2pHost.SetStreamHandler(PROTOCOL_ID, ms.msgStreamHandler)
+	ms.p2pHost.SetStreamHandler(PEER_EXCHANGE_PROTOCOL_ID, ms.receivePeerInfo)
 
-	ms.p2pHost.SetStreamHandler(PEER_EXCHANGE_PROTOCOL_ID, func(stream network.Stream) {
-		ms.receivePeerInfo(stream)
-		stream.Close()
-	})
+	if useMdnsPeerDiscovery {
+		ms.setupMdns()
+	} else {
+		ms.setupDht(bootPeers)
+	}
 
+	return ms
+}
+
+func (ms *P2PMessageService) setupMdns() {
 	// Since the mdns service could trigger a call to  `HandlePeerFound` at any time once started
 	// We want to start mdns after the message service has been fully constructed
-	if useMdnsPeerDiscovery {
-		mdns := mdns.NewMdnsService(host, "", ms)
+	ms.mdns = mdns.NewMdnsService(ms.p2pHost, "", ms)
+	err := ms.mdns.Start()
+	ms.checkError(err)
+}
+
+func (ms *P2PMessageService) setupDht(bootPeers []string) {
+	ctx := context.Background()
+	var options []dht.Option
+	if len(bootPeers) == 0 {
+		options = append(options, dht.Mode(dht.ModeServer)) // Server mode: allows other peers to connect to this node
+	}
+	kademliaDHT, err := dht.New(ctx, ms.p2pHost, options...)
+	ms.checkError(err)
+	ms.dht = kademliaDHT
 
-		ms.mdns = mdns
+	// Print out my own peerInfo
+	peerInfo := peer.AddrInfo{
+		ID:    ms.p2pHost.ID(),
+		Addrs: ms.p2pHost.Addrs(),
+	}
+	addrs, err := peer.AddrInfoToP2pAddrs(&peerInfo)
+	ms.checkError(err)
```

Comment:
Completely outside the scope of this PR but we should probably consider some kind of proper error handling for the p2p message service. Right now we just `panic`. I'll create an issue for this.

Reply:
Yes that's a good call. All messaging errors should not cause `panic`